service/matching/matcher.go

```diff
@@ -409,22 +507,22 @@ func (tm *TaskMatcher) poll(ctx context.Context, pollMetadata *pollMetadata, que
 	select {
 	case <-ctx.Done():
 		tm.metricsHandler.Counter(metrics.PollTimeoutPerTaskQueueCounter.GetMetricName()).Record(1)
-		return nil, errNoTasks
+		return nil, false, errNoTasks
 	case task := <-taskC:
 		if task.responseC != nil {
 			tm.metricsHandler.Counter(metrics.PollSuccessWithSyncPerTaskQueueCounter.GetMetricName()).Record(1)
 		}
 		tm.metricsHandler.Counter(metrics.PollSuccessPerTaskQueueCounter.GetMetricName()).Record(1)
-		return task, nil
+		return task, false, nil
 	case task := <-queryTaskC:
 		tm.metricsHandler.Counter(metrics.PollSuccessWithSyncPerTaskQueueCounter.GetMetricName()).Record(1)
 		tm.metricsHandler.Counter(metrics.PollSuccessPerTaskQueueCounter.GetMetricName()).Record(1)
-		return task, nil
+		return task, false, nil
 	}
 }
 
 func (tm *TaskMatcher) fwdrPollReqTokenC() <-chan *ForwarderReqToken {
-	if tm.fwdr == nil {
+	if tm.fwdr == nil || !tm.isBacklogNegligible() {
```

Comment:
it feels a little weird to hide this here. I think I'd prefer if it was handed at the sites that call this, even if it's more verbose. in that case, maybe it makes sense to put the `tm.fwdr == nil` case there also and get rid of this?